### DISCUSSION TITLE
How to add a stub that only loaded when `Codebase::$php_major_version` >= 8?

### DISCUSSION THREAD

**Discussion by fluffycondor** [vimeo/psalm#8327] [Q&A] [answered]
After submitting PR #8312 I noticed that it actually works not as I expected.
Looks like now Psalm thinks DatePeriod implements IteratorAggregate even on PHP 7, but the object's stub was placed in Php80.phpstub.
I made a test that makes the issue obvious:
````php
    /**
     * @return iterable<string,array{string,error_message:string,1?:string[],2?:bool,3?:string}>
     */
    public function providerInvalidCodeParse(): iterable
    {
        yield 'DatePeriod implements only Traversable on PHP 7' => [
            '<?php

            /**
             * @param mixed $value
             * @param class-string<T> $type
             * @template T
             * @psalm-assert T $value
             */
            function assertInstanceOf($value, string $type): void {
                // some code
            }

            $period = new DatePeriod("R4/2012-07-01T00:00:00Z/P7D");

            assertInstanceOf($period, IteratorAggregate::class);
            assertInstanceOf($period, stdClass::class);',
            'error_message' => 'blahblah',
            'error_levels' => [],
            'strict_mode' => false,
            'php_version' => '7.3',
        ];
    }
````
And the failure message looks like this:

> 1) Psalm\Tests\CoreStubsTest::testInvalidCode with data set "DatePeriod implements only Traversable on PHP 7" ('<?php\n\n            /**\n   ...lass);', 'blahblah', array(), false, '7.3')
> Failed asserting that exception message 'DocblockTypeContradiction - src/somefile.php:16:13 - Cannot resolve types for $period - docblock-defined type DatePeriod<"R4/2012-07-01T00:00:00Z/P7D">&IteratorAggregate does not contain stdClass' matches '/\bblahblah\b/'.

Woah, it definitely shouldn't be `DatePeriod<string>&IteratorAggregate` on PHP 7.3.
I investigated it a bit and have found that all `Php80.phpstub` entries are under `$codebase->php_major_version >= 8` logic, so looks like it shouldn't be loaded for `'php_version' => '7.3'`, but perhaps it is?..
Could it be that `@psalm-assert` relies on the language version itself rather than was set by `ProjectAnalyzer::setPhpVersion()`?
Also there is a suspicious part: while `Scanner::scanFilePaths()` purges `Scanner::$files_to_scan`, it doesn't do that for `Scanner::$files_to_deep_scan`.
https://github.com/vimeo/psalm/blob/4.x/src/Psalm/Internal/Codebase/Scanner.php#L326

Any insights are appreciated.

**Comment by orklah** [comment]
Psalm types are a combination between stubs, callmap definitions, reflection and plugins definitions. It's sometimes hard to exactly get what it does, especially when running with a different php version than the analysis. Maybe you ran from a php8 environment and reflection was prioritized instead of the stub?

Do you also reproduce using a PHP7 env?

**Reply by AndrolGenhald** [reply]
> Maybe you ran from a php8 environment and reflection was prioritized instead of the stub?

That's a good guess, very well could be what's going on. Yet another reason to work toward using reflection less if that's the case.

**Comment by AndrolGenhald** [comment]
Wow, nice find. I'll try to find some time to look at it tomorrow, that's definitely a problem if Php80.phpstub is leaking into PHP 7 projects.

**Comment by AndrolGenhald** [comment, marked as answer]
This is actually a non-issue. `IteratorAggregate` is an interface, and `DatePeriod` is extensible, so asserting that a `DatePeriod` implements `IteratorAggregate` in PHP 7 is [totally](https://psalm.dev/r/8a1362df21) [fine](https://3v4l.org/Pf4Nb).

**Reply by fluffycondor** [reply]
Thank you for the investigation!
Then I don't know how to write a test set that ensures DatePeriod only implements Traversable on PHP7 and IteratorAggregate on PHP8 :(
Also If I try to use `get_class()` in your snippet, Psalm can't detect that this condition will be never fulfilled:
https://psalm.dev/r/4250382f58

**Reply by AndrolGenhald** [reply]
> Then I don't know how to write a test set that ensures DatePeriod only implements Traversable on PHP7 and IteratorAggregate on PHP8 :(

Does just calling `->getIterator()` and making sure it has an error work?

> Also If I try to use get_class() in your snippet, Psalm can't detect that this condition will be never fulfilled

Yeah, that's a more general problem: https://psalm.dev/r/90b9310969
You can open a separate issue for that if you like, it might not be too difficult to fix if you can figure out where it's supposed to happen. Just need to check `TNamedObject::$definite_class` and disallow intersections with interfaces that aren't directly implemented.

**Reply by fluffycondor** [reply]
> Does just calling `->getIterator()` and making sure it has an error work?

Yes, this should work. Thank you!

> it might not be too difficult to fix

It can be a more complicated problem. I don't see any signs that using `get_class()` sets `true` to `TNamedObject::$definite_class`.
